README
# /* Replaced /* Replaced /* Replaced Guzzle */ */ */ is a PHP framework for building RESTful web service /* Replaced /* Replaced /* Replaced client */ */ */s

Most web service /* Replaced /* Replaced /* Replaced client */ */ */s follow a specific pattern: create a /* Replaced /* Replaced /* Replaced client */ */ */ class, create methods for each action that can be taken on the API, create a cURL handle to transfer an HTTP request to the /* Replaced /* Replaced /* Replaced client */ */ */, and return the response.  You've probably had to interact with an API that either doesn't have a PHP webservice /* Replaced /* Replaced /* Replaced client */ */ */ or the currently available PHP /* Replaced /* Replaced /* Replaced client */ */ */s are not up to an acceptable level of quality.  When facing these types of situations, you probably find yourself writing a webservice /* Replaced /* Replaced /* Replaced client */ */ */ that lacks advanced features such as persistent HTTP connections, parallel requests, over the wire logging, a PHP based caching forward proxy, a service configuration object, or exponential backoff.  Why would you spend the amount of time required to write those features though?  It's just a simple webservice /* Replaced /* Replaced /* Replaced client */ */ */ for just one of the many APIs you might be using.  But then you build another /* Replaced /* Replaced /* Replaced client */ */ */... and another.  Suddenly you find yourself with several web service /* Replaced /* Replaced /* Replaced client */ */ */s to maintain, each reeking of code duplication and lacking most, if not all, of the aforementioned features.  Enter /* Replaced /* Replaced /* Replaced Guzzle */ */ */.

/* Replaced /* Replaced /* Replaced Guzzle */ */ */ gives PHP developers a simple to extend framework for building robust webservice /* Replaced /* Replaced /* Replaced client */ */ */s that helps to eliminate code duplication by leveraging the patterns and solutions already coded into the framework.

## /* Replaced /* Replaced /* Replaced Guzzle */ */ */ gives PHP developers complete control over HTTP requests while utilizing HTTP/1.1 best practices.

### Control over HTTP requests

/* Replaced /* Replaced /* Replaced Guzzle */ */ */ gives PHP developers a tremendous amount of control over HTTP requests.  /* Replaced /* Replaced /* Replaced Guzzle */ */ */'s HTTP namespace provides a wrapper over the PHP libcurl bindings.  Most of the functionality implemented in the libcurl bindings has been simplified and abstracted by /* Replaced /* Replaced /* Replaced Guzzle */ */ */.  Developers who need access to cURL specific functionality that is not abstracted by /* Replaced /* Replaced /* Replaced Guzzle */ */ */ (e.g. proxies) can still add cURL handle specific behavior to /* Replaced /* Replaced /* Replaced Guzzle */ */ */ HTTP requests.

### Send HTTP requests in parallel

/* Replaced /* Replaced /* Replaced Guzzle */ */ */ helps PHP developers utilize HTTP/1.1 best practices including sending requests in parallel and utilizing persistent (keep-alive) HTTP connections.  Sending many HTTP requests serially (one at a time) can cause an unnecessary delay in a script's execution.  Each request must complete before a subsequent request can be sent.  By sending requests in parallel, a pool of HTTP requests can complete at the speed of the slowest request in the pool, significantly reducing the amount of time needed to execute multiple HTTP requests.  /* Replaced /* Replaced /* Replaced Guzzle */ */ */ provides a wrapper for the curl_multi functions in PHP.

### Managed persistent HTTP connections

Another extremely important aspect of the HTTP/1.1 protocol that is often overlooked by PHP webservice /* Replaced /* Replaced /* Replaced client */ */ */s is persistent HTTP connections.  Persistent connections allows data to be transferred between a /* Replaced /* Replaced /* Replaced client */ */ */ and server without the need to reconnect each time a subsequent request is sent, providing a significant performance boost to applications that need to send many HTTP requests to the same host.

HTTP requests and cURL handles are separate entities in /* Replaced /* Replaced /* Replaced Guzzle */ */ */.  In order for a request to get a cURL handle to transfer its message to a server, a request retrieves a cURL handle from a cURL handle factory.  The default cURL handle factory will maintain a pool of open cURL handles and return an already existent cURL handle (with a persistent HTTP connection) if available, or create a new cURL handle.

### Plugins for common HTTP request behavior

#### Over the wiring logging

Easily troubleshoot errors by logging everything that goes over the wire upstream and downstream, including redirect requests and entity bodies.

#### Truncated exponential backoff

Automatically retries HTTP requests using truncated exponential backoff.

#### PHP-based caching forward proxy

Provides a simple private transparent proxy cache that caches HTTP responses using HTTP headers to control caching behavior, cache lifetime, and supports ETag and Last-Modified based revalidation.

#### Cookie session plugin

Maintains cookies between HTTP requests.  Supports cookie version 1 and 2.

## /* Replaced /* Replaced /* Replaced Guzzle */ */ */ enables PHP developers to quickly create web service /* Replaced /* Replaced /* Replaced client */ */ */s.

PHP webservice /* Replaced /* Replaced /* Replaced client */ */ */s can be created in one of two ways in /* Replaced /* Replaced /* Replaced Guzzle */ */ */: using a concrete /* Replaced /* Replaced /* Replaced client */ */ */ or a dynamic /* Replaced /* Replaced /* Replaced client */ */ */.  Concrete /* Replaced /* Replaced /* Replaced client */ */ */s use a Command class for each action that can be taken on a webservice, while dynamic /* Replaced /* Replaced /* Replaced client */ */ */s dynamically build HTTP requests based on an XML service description.  Concrete /* Replaced /* Replaced /* Replaced client */ */ */s are more feature rich and testable than dynamic /* Replaced /* Replaced /* Replaced client */ */ */s, while dynamic /* Replaced /* Replaced /* Replaced client */ */ */s are quicker to implement.

## /* Replaced /* Replaced /* Replaced Guzzle */ */ */ provides a catalog of fully featured web service /* Replaced /* Replaced /* Replaced client */ */ */s.

/* Replaced /* Replaced /* Replaced Guzzle */ */ */ currently provides web service /* Replaced /* Replaced /* Replaced client */ */ */s for Amazon S3, Amazon SQS, Amazon SimpleDB, Amazon MWS, Cardinal Commerce, and Unfuddle.  Many more APIs are to follow, and hopefully other developers can contribute to the growing catalog of /* Replaced /* Replaced /* Replaced Guzzle */ */ */ webservice /* Replaced /* Replaced /* Replaced client */ */ */s.

## /* Replaced /* Replaced /* Replaced Guzzle */ */ */ is extremely well tested

/* Replaced /* Replaced /* Replaced Guzzle */ */ */ is used in production a mutli-million dollar e-commerce company.  /* Replaced /* Replaced /* Replaced Guzzle */ */ */ has 100% code coverage; every line of /* Replaced /* Replaced /* Replaced Guzzle */ */ */ has been tested using PHPUnit.

## Quick examples

### GET the google homepage

    <?php

    use /* Replaced /* Replaced /* Replaced Guzzle */ */ */\Http\Message\RequestFactory;

    // Use the RequestFactory to create a GET request
    $message = RequestFactory::getInstance()->newRequest('GET', 'http://www.google.com/');
    // Send the request and retrieve the response object
    $response = $message->send();
    echo $response->getStatusCode() . "\n";
    // Echo the raw HTTP request
    echo $request;
    // Echo the raw HTTP response
    echo $response;

### Clear the contents of an Amazon S3 bucket

    <?php

    use /* Replaced /* Replaced /* Replaced Guzzle */ */ */\Service\Aws\S3\Command\Bucket\ClearBucket;

    $command = new ClearBucket();
    $command->setBucket('test');
    $/* Replaced /* Replaced /* Replaced client */ */ */ = $this->getServiceBuilder()->getClient('michael.s3');
    $/* Replaced /* Replaced /* Replaced client */ */ */->execute($command);

CHANGELOG.md
# Change Log

Please refer to [UPGRADING](UPGRADING.md) guide for upgrading to a major version.

## 7.2.0 - 2020-10-10

### Added

- Support for PHP 8 [#2712](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2712), [#2715](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2715), [#2789](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2789)
- Support passing a body summarizer to the http errors middleware [#2795](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2795)

### Fixed

- Handle exceptions during response creation [#2591](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2591)
- Fix CURLOPT_ENCODING not to be overwritten [#2595](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2595)
- Make sure the Request always has a body object [#2804](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2804)

### Changed

- The `TooManyRedirectsException` has a response [#2660](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2660)
- Avoid "functions" from dependencies [#2712](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2712)

### Deprecated

- Using environment variable GUZZLE_CURL_SELECT_TIMEOUT [#2786](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2786)

## 7.1.1 - 2020-09-30

### Fixed

- Incorrect EOF detection for response body streams on Windows.

### Changed

- We dont connect curl `sink` on HEAD requests.
- Removed some PHP 5 workarounds

## 7.1.0 - 2020-09-22

### Added

- `/* Replaced Guzzle */Http\MessageFormatterInterface`

### Fixed

- Fixed issue that caused cookies with no value not to be stored.
- On redirects, we allow all safe methods like GET, HEAD and OPTIONS.
- Fixed logging on empty responses.
- Make sure MessageFormatter::format returns string

### Deprecated

- All functions in `/* Replaced Guzzle */Http` has been deprecated. Use static methods on `Utils` instead.
- `ClientInterface::getConfig()`
- `Client::getConfig()`
- `Client::__call()`
- `Utils::defaultCaBundle()`
- `CurlFactory::LOW_CURL_VERSION_NUMBER`

## 7.0.1 - 2020-06-27

* Fix multiply defined functions fatal error [#2699](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2699)

## 7.0.0 - 2020-06-27

No changes since 7.0.0-rc1.

## 7.0.0-rc1 - 2020-06-15

### Changed

* Use error level for logging errors in Middleware [#2629](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2629)
* Disabled IDN support by default and require ext-intl to use it [#2675](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2675)

## 7.0.0-beta2 - 2020-05-25

### Added

* Using `Utils` class instead of functions in the `/* Replaced Guzzle */Http` namespace. [#2546](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2546)
* `ClientInterface::MAJOR_VERSION` [#2583](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2583)

### Changed

* Avoid the `getenv` function when unsafe [#2531](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2531)
* Added real /* Replaced client */ methods [#2529](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2529)
* Avoid functions due to global install conflicts [#2546](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2546)
* Use Symfony intl-idn polyfill [#2550](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2550)
* Adding methods for HTTP verbs like `Client::get()`, `Client::head()`, `Client::patch()` etc [#2529](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2529)
* `ConnectException` extends `TransferException` [#2541](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2541)
* Updated the default User Agent to "/* Replaced Guzzle */Http/7" [#2654](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2654)

### Fixed

* Various intl icu issues [#2626](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2626)

### Removed

* Pool option `pool_size` [#2528](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2528)

## 7.0.0-beta1 - 2019-12-30

The diff might look very big but 95% of /* Replaced Guzzle */ users will be able to upgrade without modification.
Please see [the upgrade document](UPGRADING.md) that describes all BC breaking changes.

### Added

* Implement PSR-18 and dropped PHP 5 support [#2421](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2421) [#2474](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2474)
* PHP 7 types [#2442](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2442) [#2449](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2449) [#2466](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2466) [#2497](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2497) [#2499](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2499)
* IDN support for redirects [2424](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2424)

### Changed

* Dont allow passing null as third argument to `BadResponseException::__construct()` [#2427](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2427)
* Use SAPI constant instead of method call [#2450](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2450)
* Use native function invocation [#2444](https://github.com//* Replaced guzzle *///* Replaced guzzle *//pull/2444)

src/Client.php
<?php

namespace /* Replaced Guzzle */Http;

use /* Replaced Guzzle */Http\Cookie\CookieJar;
use /* Replaced Guzzle */Http\Exception\/* Replaced Guzzle */Exception;
use /* Replaced Guzzle */Http\Exception\InvalidArgumentException;
use /* Replaced Guzzle */Http\Promise as P;
use /* Replaced Guzzle */Http\Promise\PromiseInterface;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\UriInterface;

/**
 * @final
 */
class Client implements ClientInterface, \Psr\Http\Client\ClientInterface
{
    use ClientTrait;

    /**
     * @var array Default request options
     */
    private $config;

    /**
     * Clients accept an array of constructor parameters.
     *
     * Here's an example of creating a /* Replaced client */ using a base_uri and an array of
     * default request options to apply to each request:
     *
     *     $/* Replaced client */ = new Client([
     *         'base_uri'        => 'http://www.foo.com/1.0/',
     *         'timeout'         => 0,
     *         'allow_redirects' => false,
     *         'proxy'           => '192.168.16.1:10'
     *     ]);
     *
     * Client configuration settings include the following options:
     *
     * - handler: (callable) Function that transfers HTTP requests over the
     *   wire. The function is called with a /* Replaced Psr7 */\Http\Message\RequestInterface
     *   and array of transfer options, and must return a
     *   /* Replaced Guzzle */Http\Promise\PromiseInterface that is fulfilled with a
     *   /* Replaced Psr7 */\Http\Message\ResponseInterface on success.
     *   If no handler is provided, a default handler will be created
     *   that enables all of the request options below by attaching all of the
     *   default middleware to the handler.
     * - base_uri: (string|UriInterface) Base URI of the /* Replaced client */ that is merged
     *   into relative URIs. Can be a string or instance of UriInterface.
     * - **: any request option
     *
     * @param array $config Client configuration settings.
     *
     * @see \/* Replaced Guzzle */Http\RequestOptions for a list of available request options.
     */
    public function __construct(array $config = [])
    {
        if (!isset($config['handler'])) {
            $config['handler'] = HandlerStack::create();
        } elseif (!\is_callable($config['handler'])) {
            throw new InvalidArgumentException('handler must be a callable');
        }

        // Convert the base_uri to a UriInterface
        if (isset($config['base_uri'])) {
            $config['base_uri'] = /* Replaced Psr7 */\Utils::uriFor($config['base_uri']);
        }

        $this->configureDefaults($config);
    }

    /**
     * @param string $method
     * @param array  $args
     *
     * @return PromiseInterface|ResponseInterface
     *
     * @deprecated Client::__call will be removed in /* Replaced guzzle */http//* Replaced guzzle */:8.0.
     */
    public function __call($method, $args)
    {
        if (\count($args) < 1) {
            throw new InvalidArgumentException('Magic request methods require a URI and optional options array');
        }

        $uri = $args[0];
        $opts = $args[1] ?? [];

        return \substr($method, -5) === 'Async'
            ? $this->requestAsync(\substr($method, 0, -5), $uri, $opts)
            : $this->request($method, $uri, $opts);
    }

    /**
     * Asynchronously send an HTTP request.
     *
     * @param array $options Request options to apply to the given
     *                       request and to the transfer. See \/* Replaced Guzzle */Http\RequestOptions.
     */
    public function sendAsync(RequestInterface $request, array $options = []): PromiseInterface
    {
        // Merge the base URI into the request URI if needed.
        $options = $this->prepareDefaults($options);

        return $this->transfer(
            $request->withUri($this->buildUri($request->getUri(), $options), $request->hasHeader('Host')),
            $options
        );
    }

    /**
     * Send an HTTP request.
     *
     * @param array $options Request options to apply to the given
     *                       request and to the transfer. See \/* Replaced Guzzle */Http\RequestOptions.
     *
     * @throws /* Replaced Guzzle */Exception
     */
    public function send(RequestInterface $request, array $options = []): ResponseInterface
    {
        $options[RequestOptions::SYNCHRONOUS] = true;
        return $this->sendAsync($request, $options)->wait();
    }

    /**
     * The HttpClient PSR (PSR-18) specify this method.
     *
     * @inheritDoc
     */
    public function sendRequest(RequestInterface $request): ResponseInterface
    {
        $options[RequestOptions::SYNCHRONOUS] = true;
        $options[RequestOptions::ALLOW_REDIRECTS] = false;
        $options[RequestOptions::HTTP_ERRORS] = false;

        return $this->sendAsync($request, $options)->wait();
    }

    /**
     * Create and send an asynchronous HTTP request.
     *
     * Use an absolute path to override the base path of the /* Replaced client */, or a
     * relative path to append to the base path of the /* Replaced client */. The URL can
     * contain the query string as well. Use an array to provide a URL
     * template and additional variables to use in the URL template expansion.
     *
     * @param string              $method  HTTP method
     * @param string|UriInterface $uri     URI object or string.
     * @param array               $options Request options to apply. See \/* Replaced Guzzle */Http\RequestOptions.
     */
    public function requestAsync(string $method, $uri = '', array $options = []): PromiseInterface
    {
        $options = $this->prepareDefaults($options);
        // Remove request modifying parameter because it can be done up-front.
        $headers = $options['headers'] ?? [];
        $body = $options['body'] ?? null;
        $version = $options['version'] ?? '1.1';
        // Merge the URI into the base URI.
        $uri = $this->buildUri(/* Replaced Psr7 */\Utils::uriFor($uri), $options);
        if (\is_array($body)) {
            throw $this->invalidBody();
        }
        $request = new /* Replaced Psr7 */\Request($method, $uri, $headers, $body, $version);
        // Remove the option so that they are not doubly-applied.
        unset($options['headers'], $options['body'], $options['version']);

        return $this->transfer($request, $options);
    }

    /**
     * Create and send an HTTP request.
     *
     * Use an absolute path to override the base path of the /* Replaced client */, or a
     * relative path to append to the base path of the /* Replaced client */. The URL can
     * contain the query string as well.
     *
     * @param string              $method  HTTP method.
     * @param string|UriInterface $uri     URI object or string.
     * @param array               $options Request options to apply. See \/* Replaced Guzzle */Http\RequestOptions.
     *
     * @throws /* Replaced Guzzle */Exception
     */
    public function request(string $method, $uri = '', array $options = []): ResponseInterface
    {
        $options[RequestOptions::SYNCHRONOUS] = true;
        return $this->requestAsync($method, $uri, $options)->wait();
    }

    /**
     * Get a /* Replaced client */ configuration option.
     *
     * These options include default request options of the /* Replaced client */, a "handler"
     * (if utilized by the concrete /* Replaced client */), and a "base_uri" if utilized by
     * the concrete /* Replaced client */.
     *
     * @param string|null $option The config option to retrieve.
     *
     * @return mixed
     *
     * @deprecated Client::getConfig will be removed in /* Replaced guzzle */http//* Replaced guzzle */:8.0.
     */
    public function getConfig(?string $option = null)
    {
        return $option === null
            ? $this->config
            : (isset($this->config[$option]) ? $this->config[$option] : null);
    }

    private function buildUri(UriInterface $uri, array $config): UriInterface
    {
        if (isset($config['base_uri'])) {
            $uri = /* Replaced Psr7 */\UriResolver::resolve(/* Replaced Psr7 */\Utils::uriFor($config['base_uri']), $uri);
        }

        if (isset($config['idn_conversion']) && ($config['idn_conversion'] !== false)) {
            $idnOptions = ($config['idn_conversion'] === true) ? \IDNA_DEFAULT : $config['idn_conversion'];
            $uri = Utils::idnUriConvert($uri, $idnOptions);
        }

        return $uri->getScheme() === '' && $uri->getHost() !== '' ? $uri->withScheme('http') : $uri;
    }

    /**
     * Configures the default options for a /* Replaced client */.
     */
    private function configureDefaults(array $config): void
    {
        $defaults = [
            'allow_redirects' => RedirectMiddleware::$defaultSettings,
            'http_errors'     => true,
            'decode_content'  => true,
            'verify'          => true,
            'cookies'         => false,
            'idn_conversion'  => false,
        ];

        // Use the standard Linux HTTP_PROXY and HTTPS_PROXY if set.

        // We can only trust the HTTP_PROXY environment variable in a CLI
        // process due to the fact that PHP has no reliable mechanism to
        // get environment variables that start with "HTTP_".
        if (\PHP_SAPI === 'cli' && ($proxy = Utils::getenv('HTTP_PROXY'))) {
            $defaults['proxy']['http'] = $proxy;
        }

        if ($proxy = Utils::getenv('HTTPS_PROXY')) {
            $defaults['proxy']['https'] = $proxy;
        }

        if ($noProxy = Utils::getenv('NO_PROXY')) {
            $cleanedNoProxy = \str_replace(' ', '', $noProxy);
            $defaults['proxy']['no'] = \explode(',', $cleanedNoProxy);
        }

        $this->config = $config + $defaults;

        if (!empty($config['cookies']) && $config['cookies'] === true) {
            $this->config['cookies'] = new CookieJar();
        }

        // Add the default user-agent header.
        if (!isset($this->config['headers'])) {
            $this->config['headers'] = ['User-Agent' => Utils::defaultUserAgent()];
        } else {
            // Add the User-Agent header if one was not already set.
            foreach (\array_keys($this->config['headers']) as $name) {
                if (\strtolower($name) === 'user-agent') {
                    return;
                }
            }
            $this->config['headers']['User-Agent'] = Utils::defaultUserAgent();
        }
    }

    /**
     * Merges default options into the array.
     *
     * @param array $options Options to modify by reference
     */
    private function prepareDefaults(array $options): array
    {
        $defaults = $this->config;

        if (!empty($defaults['headers'])) {
            // Default headers are only added if they are not present.
            $defaults['_conditional'] = $defaults['headers'];
            unset($defaults['headers']);
        }

        // Special handling for headers is required as they are added as
        // conditional headers and as headers passed to a request ctor.
        if (\array_key_exists('headers', $options)) {
            // Allows default headers to be unset.
            if ($options['headers'] === null) {
                $defaults['_conditional'] = [];
                unset($options['headers']);
            } elseif (!\is_array($options['headers'])) {
                throw new InvalidArgumentException('headers must be an array');
            }
        }

        // Shallow merge defaults underneath options.
        $result = $options + $defaults;

        // Remove null values.
        foreach ($result as $k => $v) {
            if ($v === null) {
                unset($result[$k]);
            }
        }

        return $result;
    }

    /**
     * Transfers the given request and applies request options.
     *
     * The URI of the request is not modified and the request options are used
     * as-is without merging in default options.
     *
     * @param array $options See \/* Replaced Guzzle */Http\RequestOptions.
     */
    private function transfer(RequestInterface $request, array $options): PromiseInterface
    {
        $request = $this->applyOptions($request, $options);
        /** @var HandlerStack $handler */
        $handler = $options['handler'];

        try {
            return P\Create::promiseFor($handler($request, $options));
        } catch (\Exception $e) {
            return P\Create::rejectionFor($e);
        }
    }

    /**
     * Applies the array of request options to a request.
     */
    private function applyOptions(RequestInterface $request, array &$options): RequestInterface
    {
        $modify = [
            'set_headers' => [],
        ];

        if (isset($options['headers'])) {
            $modify['set_headers'] = $options['headers'];
            unset($options['headers']);
        }

        if (isset($options['form_params'])) {
            if (isset($options['multipart'])) {
                throw new InvalidArgumentException('You cannot use '
                    . 'form_params and multipart at the same time. Use the '
                    . 'form_params option if you want to send application/'
                    . 'x-www-form-urlencoded requests, and the multipart '
                    . 'option to send multipart/form-data requests.');
            }
            $options['body'] = \http_build_query($options['form_params'], '', '&');
            unset($options['form_params']);
            // Ensure that we don't have the header in different case and set the new value.
            $options['_conditional'] = /* Replaced Psr7 */\Utils::caselessRemove(['Content-Type'], $options['_conditional']);
            $options['_conditional']['Content-Type'] = 'application/x-www-form-urlencoded';
        }

        if (isset($options['multipart'])) {
            $options['body'] = new /* Replaced Psr7 */\MultipartStream($options['multipart']);
            unset($options['multipart']);
        }

        if (isset($options['json'])) {
            $options['body'] = Utils::jsonEncode($options['json']);
            unset($options['json']);
            // Ensure that we don't have the header in different case and set the new value.
            $options['_conditional'] = /* Replaced Psr7 */\Utils::caselessRemove(['Content-Type'], $options['_conditional']);
            $options['_conditional']['Content-Type'] = 'application/json';
        }

        if (!empty($options['decode_content'])
            && $options['decode_content'] !== true
        ) {
            // Ensure that we don't have the header in different case and set the new value.
            $options['_conditional'] = /* Replaced Psr7 */\Utils::caselessRemove(['Accept-Encoding'], $options['_conditional']);
            $modify['set_headers']['Accept-Encoding'] = $options['decode_content'];
        }

        if (isset($options['body'])) {
            if (\is_array($options['body'])) {
                throw $this->invalidBody();
            }
            $modify['body'] = /* Replaced Psr7 */\Utils::streamFor($options['body']);
            unset($options['body']);
        }

        if (!empty($options['auth']) && \is_array($options['auth'])) {
            $value = $options['auth'];
            $type = isset($value[2]) ? \strtolower($value[2]) : 'basic';
            switch ($type) {
                case 'basic':
                    // Ensure that we don't have the header in different case and set the new value.
                    $modify['set_headers'] = /* Replaced Psr7 */\Utils::caselessRemove(['Authorization'], $modify['set_headers']);
                    $modify['set_headers']['Authorization'] = 'Basic '
                        . \base64_encode("$value[0]:$value[1]");
                    break;
                case 'digest':
                    // @todo: Do not rely on curl
                    $options['curl'][\CURLOPT_HTTPAUTH] = \CURLAUTH_DIGEST;
                    $options['curl'][\CURLOPT_USERPWD] = "$value[0]:$value[1]";
                    break;
                case 'ntlm':
                    $options['curl'][\CURLOPT_HTTPAUTH] = \CURLAUTH_NTLM;
                    $options['curl'][\CURLOPT_USERPWD] = "$value[0]:$value[1]";
                    break;
            }
        }

        if (isset($options['query'])) {
            $value = $options['query'];
            if (\is_array($value)) {
                $value = \http_build_query($value, '', '&', \PHP_QUERY_RFC3986);
            }
            if (!\is_string($value)) {
                throw new InvalidArgumentException('query must be a string or array');
            }
            $modify['query'] = $value;
            unset($options['query']);
        }

        // Ensure that sink is not an invalid value.
        if (isset($options['sink'])) {
            // TODO: Add more sink validation?
            if (\is_bool($options['sink'])) {
                throw new InvalidArgumentException('sink must not be a boolean');
            }
        }

        $request = /* Replaced Psr7 */\Utils::modifyRequest($request, $modify);
        if ($request->getBody() instanceof /* Replaced Psr7 */\MultipartStream) {
            // Use a multipart/form-data POST if a Content-Type is not set.
            // Ensure that we don't have the header in different case and set the new value.
            $options['_conditional'] = /* Replaced Psr7 */\Utils::caselessRemove(['Content-Type'], $options['_conditional']);
            $options['_conditional']['Content-Type'] = 'multipart/form-data; boundary='
                . $request->getBody()->getBoundary();
        }

        // Merge in conditional headers if they are not present.
        if (isset($options['_conditional'])) {
            // Build up the changes so it's in a single clone of the message.
            $modify = [];
            foreach ($options['_conditional'] as $k => $v) {
                if (!$request->hasHeader($k)) {
                    $modify['set_headers'][$k] = $v;
                }
            }
            $request = /* Replaced Psr7 */\Utils::modifyRequest($request, $modify);
            // Don't pass this internal value along to middleware/handlers.
            unset($options['_conditional']);
        }

        return $request;
    }

    /**
     * Return an InvalidArgumentException with pre-set message.
     */
    private function invalidBody(): InvalidArgumentException
    {
        return new InvalidArgumentException('Passing in the "body" request '
            . 'option as an array to send a request is not supported. '
            . 'Please use the "form_params" request option to send a '
            . 'application/x-www-form-urlencoded request, or the "multipart" '
            . 'request option to send a multipart/form-data request.');
    }
}
